### PR TITLE
Adrian 

### PR DESCRIPTION
**Pull request by adrianRoussev**

### CODE REVIEW

File: lib/data.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
+require './lib/dmv'
+require './lib/facility'
+require './lib/vehicle'
+require './lib/dmv_data_service'
+require 'json'
+
+class Create
+attr_reader :facilities, :vehicles
```
Comment (KatBrandt) [inline lib/data.rb:8]: Remember to indent this line

File: lib/data.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                            facility.update({:address => "#{hash["address"]}, #{hash["city"]}, #{hash["state"]} #{hash["zip"]}"})
+                        else
+                            facility.update({:address => facility.fetch(key)})
+                        end 
+                elsif   key.to_s.include?('title') ||  key.to_s.include?('name') 
+                    facility.update ({:name => facility.fetch(key)})
+        
+                elsif   key.to_s.include?('phone') 
+                        facility.update({:phone => facility.fetch(key)}) 
+                end
```
Comment (KatBrandt) [inline lib/data.rb:34]: Yeah, I like the idea of creating a method like this where we can input any facility list and if the keys are different, we just add that new key into this method.

File: lib/data.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    
+            end
+            facility_new = Facility.new(facility)
+                @facilities << facility_new
+    end
+        @facilities
+    end
+
+
+def vehicle_factory(vehicle_list)
```
Comment (KatBrandt) [inline lib/data.rb:44]: watch that indentation

File: lib/data.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                vehicle[:make] =  vehicle[element]
+            elsif element.to_s.include?('model') && element.to_s.include?('year') == false
+                vehicle[:model] =  vehicle[element]
+            elsif element.to_s.include?('year')
+                vehicle[:year] =  vehicle[element]
+            else
+                vehicle.delete(element)
+            end 
+        end
+            vehicle.delete_if{|k,v| k != :vin && k != :make && k != :model && k != :year}
```
Comment (KatBrandt) [inline lib/data.rb:60]: There's actually no need to delete extra keys. The Vehicle model is set up to look for specific keys in the hash and ignore any other keys. This is good because if we needed to delete keys and there were 500 keys in a hash, this would take up a lot of processing power!

File: lib/data.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            @vehicle_new = Vehicle.new(vehicle)
+            @vehicles<<@vehicle_new
+    end
+        @vehicles
+    end
+end
```
Comment (KatBrandt) [inline lib/data.rb:66]: Overall I like how you approached this class. I would think about Single Responsibility Principle though. Would it make more sense to have a CreateVehicles class and a CreateFactories class? The more modular we can get with methods and classes, the easier it is to read and debug and add onto later.

File: lib/data.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
+require './lib/dmv'
```
Comment (KatBrandt) [inline lib/data.rb:1]: We only need these require statements in the test file. 
Also, convention is that the name of the file should match the class name, so this file should be create.rb

File: lib/facility.rb [hunk 7]
```diff
@@ -1,14 +1,65 @@
...
 
-  def add_services(service)
+  def add_service(service)
     @services << service
   end
+
+  
+  def register_vehicle(vehicle)
+    vehicle.change_plate
+    vehicle.change_date
```
Comment (KatBrandt) [inline lib/facility.rb:27]: I like that you put these methods in the vehicle class. That seems like the responsibility of the Vehicle class.

File: lib/facility.rb [hunk 8]
```diff
@@ -1,14 +1,65 @@
...
+    @registered_vehicles << vehicle
+    if vehicle.antique?
+      @collected_fees += 25
+    elsif vehicle.electric_vehicle?
+      @collected_fees += 200
+    else
+      @collected_fees += 100
+    end
```
Comment (KatBrandt) [inline lib/facility.rb:35]: Good use of flow control

File: lib/facility.rb [hunk 9]
```diff
@@ -1,14 +1,65 @@
...
+    
+  end
+  
+  def administer_written_test(registrant)
+    if @services.include?('Written Test') == true && registrant.permit? ==true && registrant.age >=16
```
Comment (KatBrandt) [inline lib/facility.rb:40]: Nice job with your boolean logic. One thing you can omit to make this cleaner, is if something already returns a boolean, we don't need to also do a comparison with a boolean. So line 40 could read `if @services.include?('Written Test') && registrant.permit? && registrant.age >=16`

File: spec/facility_spec.rb [hunk 10]
```diff
@@ -23,4 +24,54 @@
...
+      bolt = Vehicle.new({vin: '987654321abcdefgh', year: 2019, make: 'Chevrolet', model: 'Bolt', engine: :ev} )
+camaro = Vehicle.new({vin: '1a2b3c4d5e6f', year: 1969, make: 'Chevrolet', model: 'Camaro', engine: :ice} )
+      @facility.register_vehicle(bolt)
+      @facility.register_vehicle(camaro)
+      expect(@facility.registered_vehicles).to eq([bolt, camaro])
+      expect(@facility.collected_fees).to eq(225)
+      expect(bolt.plate_type).to eq(:ev)
+      expect(bolt.registration_date).to eq(Time.new)
+    end
+    end
```
Comment (KatBrandt) [inline spec/facility_spec.rb:39]: indentation is off here

File: lib/registrant.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
+require 'pry'
+class Registrant
+    attr_reader :name, :age, :permit, :licence_data, :written, :licence, :renewed
```
Comment (KatBrandt) [inline lib/registrant.rb:3]: We only want attr_readers defined for instance variables that we want to expose methods for. So we only want `attr_reader :name, :age, :license_data`. All the other methods are defined below so they should not go in the attr_reader.

File: lib/registrant.rb [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    def permit?
+        @permit
+    end
+
+    def earn_permit
+        @permit = true
+    end
+
+    def write_test
+    @licence_data[:written] = true
```
Comment (KatBrandt) [inline lib/registrant.rb:23]: indent

File: spec/registrant_spec.rb [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+        :renewed => false,
+        })
+    end
+  end
+
+  describe 'permit' do
+    it 'can earn permit' do
+      expect(@registrant.permit?).to eq(false)
+      @registrant.earn_permit
+      expect(@registrant.permit?).to eq(true)
```
Comment (KatBrandt) [inline spec/registrant_spec.rb:24]: Good job showing the default value, then the effects of the earn_permit method.

File: spec/registrant_spec.rb [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+            expect(@registrant.test_written?).to eq(false)
+            @registrant.write_test
+            expect(@registrant.test_written?).to eq(true)
+            expect(@registrant.road_test?).to eq(false)
+            @registrant.road_test
+            expect(@registrant.road_test?).to eq(true)
+    end
+    
+    it 'can renew licence' do
+        expect(@registrant.renew_licence).to eq(true)
```
Comment (KatBrandt) [inline spec/registrant_spec.rb:39]: I like that you created all these extra helper methods and good job testing for them too. Make sure to have one test (one it block) for every method. We want to isolate the testing of each of these methods as much as possible. Helps with future readability, debugging, and adding/removing features.

**Review by KatBrandt — COMMENTED** [review]
Overall this is looking good. I know you didn't have much time to work on this, so what you have here is definitely on the right track. 
Watch out for your indentation and line breaks, want to keep those consistent and orderly as it helps with reading and writing code.
I'd also like to see more robust testing, testing all the different potential outcomes of a method and edge case testing.
